File: Problem513/solution.py
""" 
Let n be the number of nodes in the tree and h be the height of the tree

Time Complexity: O(n)
Space Complexity: O(h)

A recursive solution using dfs.

"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:

    value:int = 0
    currentLevel:int = 0

    def dfs(self, node:TreeNode, level = 0):
        if node is None:
            return
    
        if level > self.currentLevel:
            self.value = node.val
            self.currentLevel = level

        self.dfs(node.left, level+1)
        self.dfs(node.right, level+1)
    
    def findBottomLeftValue(self, root: TreeNode) -> int:
        self.value = root.val
        self.dfs(root)
        return self.value
    
sol:Solution = Solution()
root = TreeNode(1, TreeNode(2, TreeNode(4), None), TreeNode(3, TreeNode(5, TreeNode(7, None)), TreeNode(6)))
print(sol.findBottomLeftValue(root))

File: Problem513/solution.cpp
/* 
Let n be the number of nodes in the tree and h be the height of the tree

Time Complexity: O(n)
Space Complexity: O(h)

A recursive solution using dfs.

*/

#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void dfs(TreeNode* node, int level = 0){
        if(node == nullptr){return;}  
        if(level > currentLevel){
            currentLevel = level;
            value = node->val;
        }
        dfs(node->left, level+1);
        dfs(node->right, level+1);
    }

    int findBottomLeftValue(TreeNode* root) {
        value = root->val;
        dfs(root);
        return value;
    }
private:
    int value = 0;
    int currentLevel = 0;
};

int main(){

    Solution sol = Solution();
    TreeNode* root = new TreeNode(1, new TreeNode(2, new TreeNode(4), nullptr), new TreeNode(3, new TreeNode(5, new TreeNode(7), nullptr), new TreeNode(6)));

    cout << sol.findBottomLeftValue(root) << endl;

    delete root->right;
    delete root->left;
    delete root;

}